backend/aws_helpers/dynamo_db_utils/dynamo_db_utils.py

```diff
@@ -0,0 +1,198 @@
+import boto3
+from backend.aws_helpers.dynamo_db_utils.constants import ALL_DYANMODB_TABLES
+import random
+from datetime import datetime
+from typing import Union
+
+dynamodb = boto3.resource("dynamodb")
+
+
+def get_dynamo_item_by_id(table_name: str, partition_key_value: str) -> dict:
+    """
+    Get item from DynamoDB table that match all keys, raises exception if item does not exist
+
+    Args:
+        table_name (str): Name of DynamoDB table
+        partition_key_value (str): Value of partition key to match
+
+    Returns:
+        dict: One item from DynamoDB table in JSON format
+
+    Raises:
+        ValueError: If table_name is not a valid table name
+        Exception: If DynamoDB get_item call fails for any reason, e.g., item not found
+    """
+    # Validation steps
+    if table_name not in ALL_DYANMODB_TABLES.keys():
+        raise ValueError("Invalid table name: " + table_name)
+
+    # Get item
+    partition_key = ALL_DYANMODB_TABLES[table_name]["partition_key"]
+    item_key = {partition_key: partition_key_value}
+    table = dynamodb.Table(table_name)
+    response = table.get_item(Key=item_key)
+    if response.get("Item") is None:
+        raise Exception("Item not found")
+
+    return response["Item"]
+
+
+def get_dynamo_items_by_gsi(table_name: str, gsi_value: Union[int, str]) -> list[dict]:
+    """
+    Get items from DynamoDB table that match the given GSI value, returns [] if no items match
+
+    Args:
+        table_name (str): Name of DynamoDB table
+        gsi_value (int or str): Value of GSI to match
+
+    Returns:
+        list[dict]: One array containing all matching items from DynamoDB table in JSON format
+
+    Raises:
+        ValueError: If table_name is not a valid table name
+        Exception: If DynamoDB get_item call fails for any reason
+    """
+    # Validation steps
+    if table_name not in ALL_DYANMODB_TABLES.keys():
+        raise ValueError("Invalid table name: " + table_name)
+    gsi_key = ALL_DYANMODB_TABLES[table_name].get("gsi")
+    if gsi_key is None:
+        raise Exception("Table does not have a GSI")
+
+    # Make query params
+    PREFIX = "#dlp__"
+
+    # create update expression
+    key_condition_expression = f"{PREFIX}{gsi_key} = :{gsi_key}"
+
+    # create expression attribute values
+    expression_attribute_values = {f":{gsi_key}": gsi_value}
+
+    # create expression attribute names to prevent reserved words error
+    expression_attribute_names = {f"{PREFIX}{gsi_key}": gsi_key}
+
+    # Get items
+    table = dynamodb.Table(table_name)
```

Comment:
@farisdurrani you will need to use LastEvaluatedKey to ensure u are getting all items? Dynamo Db paginates query results so if u dont while loop on LastEvaluatedKey, you will only get results on the first page instead of all results?

Reply:
I missed this again. Fixed in "added LastEvaluatedKey check for gsi query". Validated by running the `print(4, getAllUserTrainspaceData("bleh"))` test